pwndbg/gdblib/hooks.py

```diff
@@ -54,43 +46,28 @@ def on_stop() -> None:
     pwndbg.gdblib.strings.update_length()
 
 
-@pwndbg.gdblib.events.stop
-@pwndbg.gdblib.events.mem_changed
-@pwndbg.gdblib.events.reg_changed
-def memoize_on_stop() -> None:
-    reset_on_stop._reset()
-
-
-@pwndbg.gdblib.events.before_prompt
-def memoize_before_prompt() -> None:
-    reset_on_prompt._reset()
-
-
-@pwndbg.gdblib.events.cont
-def memoize_on_cont() -> None:
-    reset_on_cont._reset()
-
-
-@pwndbg.gdblib.events.new_objfile
-def memoize_on_new_objfile() -> None:
-    reset_on_objfile._reset()
-
-
-@pwndbg.gdblib.events.start
-def memoize_on_start() -> None:
-    while_running._start_caching()
-    reset_on_start._reset()
-
-
-@pwndbg.gdblib.events.exit
-def memoize_on_exit() -> None:
-    while_running._reset()
-    reset_on_exit._reset()
-
-
-@pwndbg.gdblib.events.thread
-def memoize_on_new_thread() -> None:
-    reset_on_thread._reset()
+import pwndbg.lib.cache
+
+pwndbg.lib.cache.connect_clear_caching_events(
+    (
+        (
+            "stop",
+            (
+                pwndbg.gdblib.events.stop,
+                pwndbg.gdblib.events.mem_changed,
+                pwndbg.gdblib.events.reg_changed,
+            ),
```

Comment:
I think I understand why `mem_changed` and `reg_changed` are only necessary for the `stop` event, but this could benefit from a comment. Would this statement be correct?
```
Any cache that should be cleared when the program is stopped should also be cleared if the user does an operation to modify memory or registers while the program is stopped. This isn't necessary for the other events, because there is no way for memory or registers to be changed if the program is not stopped.
```

Reply:
> because there is no way for memory or registers to be changed if the program is not stopped.

Sadly this is not entirely true since a hook can set regs/memory via gdb.execute :/